jscomp/runtime/caml_obj.res

```diff
@@ -0,0 +1,455 @@
+@ocaml.doc(" 
+   For the empty dummy object, whether it's 
+   [[]] or [{}] depends on how 
+   runtime encoding works, and will affect 
+   js polymorphic comparison(Js.(=)) (fine with caml polymoprhic comparison (Pervasives.equal))
+   In most cases, rec value comes from record/modules, 
+   whose tag is 0, we optimize that case
+")
+let update_dummy: (_, _) => unit = %raw(`function(x,y){
+  var k  
+  if(Array.isArray(y)){
+    for(k = 0; k < y.length ; ++k){
+      x[k] = y[k]
+    }
+    if(y.TAG !== undefined){
+      x.TAG = y.TAG
+    }
+  } else {
+    for (var k in y){
+      x[k] = y[k]
+    }
+  }
+}
+`)
+
+@ocaml.doc(" TODO: investigate total
+    [compare x y] returns [0] if [x] is equal to [y],
+    a negative integer if [x] is less than [y],
+    and a positive integer if [x] is greater than [y].
+    The ordering implemented by compare is compatible with the comparison
+    predicates [=], [<] and [>] defined above, with one difference on the treatment of the float value
+    [nan].
+
+    Namely, the comparison predicates treat nan as different from any other float value,
+    including itself; while compare treats [nan] as equal to itself and less than any other float value.
+    This treatment of [nan] ensures that compare defines a total ordering relation.
+    compare applied to functional values may raise Invalid_argument. compare applied to cyclic structures
+    may not terminate.
+
+    The compare function can be used as the comparison function required by the [Set.Make] and [Map.Make] functors,
+    as well as the [List.sort] and [Array.sort] functions.
+")
+let rec compare = (a: Obj.t, b: Obj.t): int =>
+  if a === b {
+    0
+  } else {
+    /* front and formoest, we do not compare function values */
+    let a_type = Js.typeof(a)
+    let b_type = Js.typeof(b)
+    switch (a_type, b_type) {
+    | ("undefined", _) => -1
+    | (_, "undefined") => 1
+    /* [a] is of type string, b can not be None,
+        [a] could be (Some (Some x)) in that case [b] could be [Some None] or [null]
+         so [b] has to be of type string or null */
+    | ("string", "string") => Pervasives.compare((Obj.magic(a): string), Obj.magic(b))
+    | ("string", _) => /* [b] could be [Some None] or [null] */
+      1
+    | (_, "string") => -1
+    | ("boolean", "boolean") => Pervasives.compare((Obj.magic(a): bool), Obj.magic(b))
+    | ("boolean", _) => 1
+    | (_, "boolean") => -1
+    | ("function", "function") => raise(Invalid_argument("compare: functional value"))
+    | ("function", _) => 1
+    | (_, "function") => -1
+    | ("bigint", "bigint")
+    | ("number", "number") =>
+      Pervasives.compare((Obj.magic(a): float), (Obj.magic(b): float))
+    | ("bigint", _) | ("number", _) =>
+      if b === Obj.repr(Js.null) || Caml_option.isNested(b) {
+        1
+      } else {
+        /* Some (Some ..) < x */
+        -1
+      } /* Integer < Block in OCaml runtime GPR #1195, except Some.. */
+    | (_, "bigint") | (_, "number") =>
```

Comment:
This too

Reply:
And this line switch back to just "number"?
```
        -1
      } /* Integer < Block in OCaml runtime GPR #1195, except Some.. */
>>>>    | (_, "bigint") | (_, "number") =>
      if a === Obj.repr(Js.null) || Caml_option.isNested(a) {
        -1
```